api/agent/pure_runner.go

```diff
@@ -425,31 +432,10 @@ func (ch *callHandle) Write(data []byte) (int, error) {
 		//If it is an detached call we just /dev/null the data coming back from the container
 		return len(data), nil
 	}
-	var err error
-	ch.headerOnce.Do(func() {
-		// WARNING: we do fetch Status and Headers without
-		// a lock below. This is a problem in agent in general, and needs
-		// to be fixed in all accessing go-routines such as protocol/http.go,
-		// protocol/json.go, agent.go, etc. In practice however, one go routine
-		// accesses them (which also compiles and writes headers), but this
-		// is fragile and needs to be fortified.
-		err = ch.enqueueMsg(&runner.RunnerMsg{
-			Body: &runner.RunnerMsg_ResultStart{
-				ResultStart: &runner.CallResultStart{
-					Meta: &runner.CallResultStart_Http{
-						Http: &runner.HttpRespMeta{
-							Headers:    ch.prepHeaders(),
-							StatusCode: int32(ch.status),
-						},
-					},
-				},
-			},
-		})
-	})
 
-	if err != nil {
-		return 0, err
-	}
+	ch.WriteHeader(ch.status)
```

Comment:
>  In the code we have in master if we see any errors during the send of RunnerMsg_ResultStart during the Write call, the Write return that error now that is not the case anymore.

pursuant to this, since we're using enqueueMsgStrict in WriteHeader then this will still bail, right? if not, it may be worth saving any errors from WriteHeader on the call handle and then checking that error here after calling WriteHeader, to block any further Write.

Reply:
probably, we need this block here (or call a new function that does this):
`
select {
	case <-ch.ctx.Done():
	        return io.EOF
	case <-ch.doneQueue:
        	return io.EOF
        default:
	}
`